src/common/heartbeatService.js

```diff
@@ -0,0 +1,104 @@
+const {
+  APPLICATION_TAG_KEY,
+  CLUSTER_TAG_KEY,
+  SERVICE_TAG_KEY,
+  SHARD_TAG_KEY,
+  COMPONENT_TAG_KEY,
+  HEART_BEAT_METRIC,
+  HEART_BEAT_INTERVAL,
+  NULL_TAG_VAL
+} = require('./constants');
+
+/**
+ * Service that periodically reports component heartbeats to Wavefront.
+ */
+class HeartbeatService {
+  /**
+   *
+   * @param {WavefrontDirectClient} wavefrontClient
+   * @param applicationTags
+   * @param {Array} components - Array of string indicates Components.
+   * @param source
+   */
+  constructor(wavefrontClient, applicationTags, components, source) {
+    this.wavefrontClient = wavefrontClient;
+    this.applicationTags = applicationTags;
+    this.source = source;
+    this.reportingIntervalSeconds = HEART_BEAT_INTERVAL;
+    this.heartbeatMetricTagsList = [];
+    this.customTagsSet = new Set();
+
+    if (!Array.isArray(components)) {
+      components = [components];
+    }
+
+    for (let component of components) {
+      let metricTags = {
+        APPLICATION_TAG_KEY: applicationTags.application,
+        CLUSTER_TAG_KEY: applicationTags.cluster || NULL_TAG_VAL,
+        SERVICE_TAG_KEY: applicationTags.service,
+        SHARD_TAG_KEY: applicationTags.shard || NULL_TAG_VAL,
+        COMPONENT_TAG_KEY: component
+      };
+
+      if (applicationTags.customTags) {
```

Comment:
If `applicationTags` was undefined or null, this line would throw a js error.

Reply:
I think applicationTags cannot be null since by default we need to specify `application` and `service`. But still, it would be a good practice to optional chaining here :)